### DISCUSSION TITLE
User certificates not from system storage on Android.

### DISCUSSION THREAD

**Discussion by ghost** [strongswan/strongswan#480] [Ideas]
Would it be possible to implement support for loading user certificates from ordinary directories in the file system rather than from the system storage?

The system storage access is restricted on many Android devices, and on Android 11 it should not even be possible (I didn't check, as I do not have Android 11) to install user certificates directly from an app, the user would have to do it from the system settings. 

I think that a "custom vpn keypair" is not the most expected by Google usage of the system storage. (I think they expect more of something bank-related.)

**Comment by tobiasbrunner** [comment]
Depending on the configuration backend, certificates and keys can be loaded from anywhere (in particular when using [vici](https://wiki.strongswan.org/projects/strongswan/wiki/vici)). Anyway, our Android app uses the KeyChain API to access client certificates and private keys.
